larky/src/main/resources/stdlib/parse.star

```diff
@@ -0,0 +1,161 @@
+# This module defines a standard interface to break Uniform Resource Locator (URL)
+# strings up in components (addressing scheme, network location, path etc.)
+
+# Direct copy from: https://github.com/python/cpython/blob/3.9/Lib/urllib/parse.py
+#
+
+load("@stdlib/larky", "larky")
+load("@stdlib//dicts", "dicts")
+load("@stdlib//builtins","builtins")
+load("@stdlib//types", "types")
+
+# Unsafe bytes to be removed per WHATWG spec
+_UNSAFE_URL_BYTES_TO_REMOVE = ['\t', '\r', '\n']
+
+MAX_CACHE_SIZE = 20
+
+uses_params = ['', 'ftp', 'hdl', 'prospero', 'http', 'imap',
+               'https', 'shttp', 'rtsp', 'rtspu', 'sip', 'sips',
+               'mms', 'sftp', 'tel']
+
+_implicit_encoding = 'ascii'
+_implicit_errors = 'strict'
+
+MAX_CACHE_SIZE = 20
+_parse_cache = {}
+
+_safe_quoters = {}
+
+# Characters valid in scheme names
+scheme_chars = ('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+-.')
+
+def _urlparse(url, scheme='', allow_fragments=True):
+    url, scheme, _coerce_result = _coerce_args(url, scheme)
+    splitresult = _urlsplit(url, scheme, allow_fragments)
+    scheme, netloc, url, query, fragment = splitresult['scheme'],splitresult['netloc'],splitresult['path'],splitresult['query'],splitresult['fragment']
+    if scheme in uses_params and ';' in url:
+        url, params = _splitparams(url)
+    else:
+        params = ''
+    result = dicts.add({'scheme': scheme}, {'netloc': netloc}, {'path': url}, {'query': query}, {'params': params}, {'fragment': fragment})
+    # ParseResult(scheme, netloc, url, params, query, fragment)
+    # return _coerce_result(result)
+    return result
+
+
+def _decode_args(args, encoding=_implicit_encoding,
+                       errors=_implicit_errors):
+    t = ()
+    for x in args:
+        if x:
+            t = t + (x.decode(encoding, errors),)
```

Comment:
what is decode? maybe we can use codecs module?

```python
load("@stdlib//codecs, "codecs")
...
t = t + (codecs.decode(x, encoding, errors),)
```

Reply:
sure. I will check this one